File: backend/routers/reveal.py
"""
Reveal-Identity router.
Recruiter or Admin — only one who clicks "Reveal Identity" button.
Returns the original filename and full candidate details.
Logs every reveal to identity_reveal_log for audit.
"""
from fastapi import APIRouter, HTTPException, Depends
from datetime import datetime, timezone
from services.supabase_client import get_supabase
from auth import require_recruiter_or_admin

router = APIRouter()


@router.post("/reveal-identity/{resume_id}")
async def reveal_identity(
    resume_id: str,
    current_user: dict = Depends(require_recruiter_or_admin),
):
    """
    Permanently reveal a candidate's identity.
    - Recruiter or Admin only (403 for all other roles)
    - Returns original_file_name stored at upload time
    - Logs reveal action in `identity_reveal_log` table
    - Marks resume as `identity_revealed = true`
    Anonymity is maintained during evaluation — this is ONLY called after ranking.
    """
    sb = get_supabase()

    # Fetch resume — retrieve PII only at reveal time
    res = sb.table("resumes")\
        .select("id, session_id, original_file_name, identity_revealed, overall_score, uploaded_by")\
        .eq("id", resume_id)\
        .single()\
        .execute()

    if not res.data:
        raise HTTPException(status_code=404, detail="Resume not found.")

    resume = res.data
    now = datetime.now(timezone.utc).isoformat()

    already_revealed = resume.get("identity_revealed", False)

    # Always log every reveal attempt (idempotent)
    try:
        sb.table("identity_reveal_log").insert({
            "resume_id": resume_id,
            "session_id": resume.get("session_id"),
            "revealed_by": current_user["id"],
            "revealed_at": now,
            "already_revealed": already_revealed,
        }).execute()
    except Exception:
        pass  # Non-fatal — don't block reveal if log insert fails

    # Mark resume as revealed
    sb.table("resumes").update({
        "identity_revealed": True,
        "identity_revealed_at": now,
        "identity_revealed_by": current_user["id"],
    }).eq("id", resume_id).execute()

    # Extract a presentable candidate name from the filename
    raw_name = resume.get("original_file_name", "Unknown")
    # Strip extension for display: "John_Smith.pdf" → "John Smith"
    display_name = raw_name
    for ext in (".pdf", ".docx", ".doc", ".txt"):
        if display_name.lower().endswith(ext):
            display_name = display_name[: -len(ext)]
    display_name = display_name.replace("_", " ").replace("-", " ").strip()
    if not display_name:
        display_name = raw_name

    return {
        "resume_id": resume_id,
        "session_id": resume.get("session_id"),
        "original_file_name": raw_name,
        "display_name": display_name,
        "overall_score": resume.get("overall_score", 0),
        "revealed_at": now,
        "revealed_by_email": current_user.get("email", ""),
        "revealed_by_role": current_user.get("role", ""),
        "already_revealed": already_revealed,
    }


@router.get("/reveal-identity/status/{resume_id}")
async def reveal_status(
    resume_id: str,
    current_user: dict = Depends(require_recruiter_or_admin),
):
    """Check if a resume's identity has already been revealed."""
    sb = get_supabase()
    res = sb.table("resumes")\
        .select("id, identity_revealed, identity_revealed_at, original_file_name")\
        .eq("id", resume_id)\
        .single()\
        .execute()

    if not res.data:
        raise HTTPException(status_code=404, detail="Resume not found.")

    r = res.data
    if r.get("identity_revealed"):
        raw = r.get("original_file_name", "")
        display = raw
        for ext in (".pdf", ".docx", ".doc", ".txt"):
            if display.lower().endswith(ext):
                display = display[: -len(ext)]
        display = display.replace("_", " ").replace("-", " ").strip() or raw
        return {
            "revealed": True,
            "original_file_name": raw,
            "display_name": display,
            "revealed_at": r.get("identity_revealed_at"),
        }
    return {"revealed": False, "original_file_name": None, "display_name": None, "revealed_at": None}

File: backend/src/routes/reveal.js
/**
 * Reveal-Identity router — mirrors Python routers/reveal.py.
 *
 * POST /api/reveal-identity/:resume_id   — permanently reveal candidate identity
 * GET  /api/reveal-identity/status/:resume_id — check if already revealed
 */
const express = require('express');
const { requireRecruiterOrAdmin } = require('../middleware/auth');
const { getSupabase } = require('../lib/supabase');

const router = express.Router();

function buildDisplayName(raw) {
  let display = raw || 'Unknown';
  for (const ext of ['.pdf', '.docx', '.doc', '.txt']) {
    if (display.toLowerCase().endsWith(ext)) {
      display = display.slice(0, -ext.length);
    }
  }
  display = display.replace(/_/g, ' ').replace(/-/g, ' ').trim();
  return display || raw;
}

// POST /api/reveal-identity/:resume_id
router.post('/:resume_id', requireRecruiterOrAdmin, async (req, res) => {
  try {
    const sb = getSupabase();
    const { resume_id } = req.params;
    const user = req.currentUser;

    const { data: resume, error } = await sb
      .from('resumes')
      .select('id, session_id, original_file_name, identity_revealed, overall_score, uploaded_by')
      .eq('id', resume_id)
      .single();

    if (error || !resume) return res.status(404).json({ detail: 'Resume not found.' });

    const now = new Date().toISOString();
    const already_revealed = resume.identity_revealed || false;

    // Always log every reveal attempt (idempotent)
    try {
      await sb.from('identity_reveal_log').insert({
        resume_id,
        session_id: resume.session_id,
        revealed_by: user.id,
        revealed_at: now,
        already_revealed,
      });
    } catch (_) {
      // Non-fatal
    }

    // Mark resume as revealed
    await sb.from('resumes').update({
      identity_revealed: true,
      identity_revealed_at: now,
      identity_revealed_by: user.id,
    }).eq('id', resume_id);

    const raw_name = resume.original_file_name || 'Unknown';
    const display_name = buildDisplayName(raw_name);

    return res.json({
      resume_id,
      session_id: resume.session_id,
      original_file_name: raw_name,
      display_name,
      overall_score: resume.overall_score || 0,
      revealed_at: now,
      revealed_by_email: user.email || '',
      revealed_by_role: user.role || '',
      already_revealed,
    });
  } catch (err) {
    return res.status(500).json({ detail: err.message });
  }
});

// GET /api/reveal-identity/status/:resume_id
router.get('/status/:resume_id', requireRecruiterOrAdmin, async (req, res) => {
  try {
    const sb = getSupabase();
    const { resume_id } = req.params;

    const { data: r, error } = await sb
      .from('resumes')
      .select('id, identity_revealed, identity_revealed_at, original_file_name')
      .eq('id', resume_id)
      .single();

    if (error || !r) return res.status(404).json({ detail: 'Resume not found.' });

    if (r.identity_revealed) {
      const raw = r.original_file_name || '';
      const display = buildDisplayName(raw);
      return res.json({
        revealed: true,
        original_file_name: raw,
        display_name: display,
        revealed_at: r.identity_revealed_at,
      });
    }

    return res.json({
      revealed: false,
      original_file_name: null,
      display_name: null,
      revealed_at: null,
    });
  } catch (err) {
    return res.status(500).json({ detail: err.message });
  }
});

module.exports = router;
